### PR TITLE
Add TCRdock structural validation (#25)

### LINKED ISSUE #25: Add structural validation of top neoepitope candidates (TCRdock + ESMFold)
## Motivation

The pipeline currently produces sequence-level predictions (MHC binding via
MHCflurry, TCR recognition via ProTCR #24). Adding structural validation of
the top candidates would confirm that predicted binders adopt a physically
plausible binding pose in the MHC groove and TCR interface — going from
sequence to 3D structure.

## Tools

Both tools are freely available without academic registration or institutional
affiliation — consistent with the pipeline's open-access principle.

| Tool | License | Use case |
|------|---------|----------|
| **TCRdock** | MIT | Full TCR-peptide-MHC ternary complex |
| **ESMFold** | MIT | MHC-peptide structure when no TCR is available |

**TCRdock** is built on a modified AlphaFold multimer, adapted specifically
for TCR:pMHC complexes. Inputs: TSV with allele, peptide, paired CDR3
sequences (from TRUST4 #24). Outputs: predicted PDB + docking geometry
parameters (distance, angle, orientation of TCR over pMHC).

**ESMFold** (Meta, MIT) is a fast protein language model-based structure
predictor. Used as a fallback when TRUST4 finds no TCR sequences — it can
still model the peptide-MHC complex alone.

**PANDORA** was considered but ruled out — it depends on MODELLER which
requires registration even for academic use.

## Pipeline position

Applied only to the top N candidates (e.g. top 10 strong binders) after
ProTCR ranking, to keep compute tractable:

```
ProTCR ranked neoepitopes
          ↓
    top N candidates
          ↓
   TCR sequences        no TCR sequences
   available?     ──────────────────────→ ESMFold (MHC-peptide)
       ↓
   TCRdock (TCR-peptide-MHC)
          ↓
   PDB structure + docking geometry score
          ↓
   Final report with embedded 3D viewer (Mol* / NGL)
```

## Report

Embed an interactive 3D structure viewer (Mol* or NGL viewer) in the HTML
report for the top structural candidates — showing the full TCR-peptide-MHC
complex where available.

## References

- TCRdock: https://github.com/phbradley/TCRdock
- ESMFold: https://github.com/facebookresearch/esm

### PR DESCRIPTION
**Pull request by Jin-HoMLee**
## Summary

Adds optional TCRdock structural validation as step 6 of the pipeline. Top MHC-binding neoepitope candidates are docked with TCR using AlphaFold via TCRdock, and the resulting 3D structures are embedded in the HTML report with an interactive Mol* viewer.

### New files
- `workflow/scripts/run_tcrdock.py` — TCRdock wrapper: selects top candidates, builds input, runs Docker, collects and relabels PDB output
- `workflow/rules/tcrdock.smk` — Snakemake rule for the TCRdock step
- `docker/Dockerfile.pipeline` — Docker image (CUDA 11.8 + cuDNN 8 + JAX 0.3.25 + AlphaFold params)
- `scripts/run_cloud_gpu.sh` — Three-phase cloud pipeline: CPU VM (steps 1–5) → GCS handoff → GPU Spot VM (TCRdock)
- `scripts/setup_tcrdock_vm.sh` — GPU VM provisioning script
- `config/tcrdock_gpu.yaml` — GPU-specific Snakemake config overrides

### Modified files
- `workflow/scripts/generate_report.py` — Interactive Mol* 3D viewer section with chain legend; PDB COMPND records for meaningful chain names
- `Snakefile` — Includes `tcrdock.smk` when `tcrdock.enabled` is set
- `config/config.yaml` — New `tcrdock.*` configuration parameters
- `README.md` — Updated pipeline diagram (7 steps), config table, output tree, project structure, citations
- `docs/google_cloud_guide.md` — New "Automated GPU Pipeline (TCRdock)" section
- `CLAUDE.md` — TCRdock Docker and PDB chain relabelling decision notes
- `docs/lab_notebook.md` — Session log for April 9–10

## Design decisions
- **Docker over conda** for TCRdock — irreconcilable cuDNN/JAX/openmm version conflicts in conda
- **PDB chain relabelling** — AlphaFold outputs a single chain; `relabel_pdb_chains()` reassigns A=MHC, B=peptide, C=TCR-α, D=TCR-β
- **GCS bucket handoff** between CPU and GPU VMs — no need for both to be up simultaneously

## Testing
- 57 unit tests passing
- Local test run (chr22 dataset) ✓
- Full cloud GPU run ✓
- Mol* chain names verified in report ✓

Closes #25

### CODE REVIEW

File: workflow/scripts/generate_report.py [hunk 1]
```diff
@@ -71,13 +72,110 @@
...
 <!DOCTYPE html>
 <html lang="en">
 <head>
   <meta charset="UTF-8"/>
   <meta name="viewport" content="width=device-width, initial-scale=1.0"/>
   <title>Splice Neoepitope Report</title>
+  <!-- Mol* molecular viewer (MIT license) — only used when TCRdock structural
+       validation is enabled (config[tcrdock][enabled]: true). -->
+  <script src="https://www.unpkg.com/molstar/build/viewer/molstar.js"></script>
+  <link rel="stylesheet" href="https://www.unpkg.com/molstar/build/viewer/molstar.css"/>
```
[the remark this answers is not in the document]
Comment (Jin-HoMLee) [reply workflow/scripts/generate_report.py]: Fixed in 6cbf9cc — Mol* assets are now only injected into the HTML when a structure section is present. Simpler approach than the on-demand loader: the template uses a `{molstar_assets}` placeholder that's filled conditionally.

File: workflow/scripts/generate_report.py [hunk 2]
```diff
@@ -254,12 +363,63 @@ def _df_to_html(df: pd.DataFrame, max_rows: int = 100) -> str:
...
+        pdb_text = pdb_path.read_text()
+    except OSError as exc:
+        log.warning("Could not read PDB file %s: %s", pdb_path, exc)
+        return "<p><em>Structure not available.</em></p>"
+
+    # Annotate with the top candidate's peptide and allele
+    top = pred_df[pred_df["binder_class"] == "strong"].sort_values("ic50_nM").head(1)
+    if top.empty:
+        peptide, allele = "top candidate", ""
+    else:
```
[the remark this answers is not in the document]
Comment (Jin-HoMLee) [reply workflow/scripts/generate_report.py]: Fixed in 92b27d1 — now filters to strong/weak binders and sorts by IC50, matching `run_tcrdock.py`'s selection logic. Fallback label changed from `"top candidate"` to `"NA"`.

File: workflow/scripts/run_tcrdock.py [hunk 3]
```diff
@@ -0,0 +1,494 @@
...
+                         (columns: peptide, allele, va_gene, vb_gene,
+                          cdr3a, cdr3b, rmsd, tcr_pdb_rmsd, docking_geometry)
+
+Usage (standalone, Linux + GPU only):
+  python run_tcrdock.py \\
+      --predictions-tsv results/predictions/local/predictions.tsv \\
+      --output-pdb results/predictions/local/tcrdock/top_candidate.pdb \\
+      --output-scores results/predictions/local/tcrdock/docking_scores.tsv \\
+      --tcrdock-dir /opt/tcrdock \\
+      --alphafold-params /opt/af_params
```
[the remark this answers is not in the document]
Comment (Jin-HoMLee) [reply workflow/scripts/run_tcrdock.py]: Fixed in 53584b9 — removed the stale `--tcrdock-dir` and `--alphafold-params` flags from the docstring.

File: workflow/scripts/run_tcrdock.py [hunk 4]
```diff
@@ -0,0 +1,494 @@
...
+
+    def _container_path(host_path: Path) -> str:
+        """Translate a host path inside output_dir to its /data/... equivalent."""
+        return "/data/" + str(host_path.relative_to(output_dir))
+
+    def _docker_run(container_cmd: list, label: str) -> None:
+        cmd = [
+            "sudo", "docker", "run", "--rm", "--gpus", "all",
+            "-v", f"{output_dir}:/data",
+            docker_image,
```
[the remark this answers is not in the document]
Comment (Jin-HoMLee) [reply workflow/scripts/run_tcrdock.py:206]: Fixed in f0ac92e — removed `sudo`. The orchestration layer (`run_cloud_gpu.sh`) runs `setup_tcrdock_vm.sh` and the pipeline in separate SSH sessions, so docker group membership is guaranteed active by the time `run_tcrdock.py` calls `docker run`.

File: workflow/rules/tcrdock.smk [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        PDB structure and docking geometry scores.
+        """
+        input:
+            predictions_tsv=rules.run_mhcflurry.output.predictions_tsv,
+        output:
+            pdb=os.path.join(
+                OUT["predictions"], "{cancer_type}", "tcrdock", "top_candidate.pdb"
+            ),
+            scores_tsv=os.path.join(
+                OUT["predictions"], "{cancer_type}", "tcrdock", "docking_scores.tsv"
```
[the remark this answers is not in the document]
Comment (Jin-HoMLee) [reply workflow/rules/tcrdock.smk:42]: No change needed — Snakemake automatically injects `os` into the workflow globals. None of the other `.smk` files in this repo import `os` either (e.g. `assemble.smk`, `filter.smk`, `predict.smk` all use `os.path.join` without importing it).

File: scripts/run_cloud_gpu.sh [hunk 6]
```diff
@@ -0,0 +1,503 @@
...
+ZONE="europe-west1-b"
+MACHINE_TYPE="n1-standard-4"
+ACCELERATOR="type=nvidia-tesla-t4,count=1"
+DISK_SIZE="100GB"
+IMAGE_FAMILY="common-cu128-ubuntu-2204-nvidia-570"  # CUDA 12.8 pre-installed
+IMAGE_PROJECT="deeplearning-platform-release"
+REPO_URL="https://github.com/Jin-HoMLee/splice-neoepitope-pipeline.git"
+ORCH_VM="neoepitope-orchestrator"
+
+GCS_BUCKET="tcrdock-handoff"
```
[the remark this answers is not in the document]
Comment (Jin-HoMLee) [reply scripts/run_cloud_gpu.sh]: Fixed in 3f7b6c6 — bucket name is now derived from the active GCP project ID: `$(gcloud config get-value project)-tcrdock-handoff`.

File: docker/Dockerfile.pipeline [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    dm-tree==0.1.8 \
+    immutabledict==2.0.0 \
+    jax==0.3.25 \
+    "https://storage.googleapis.com/jax-releases/cuda11/jaxlib-0.3.25+cuda11.cudnn805-cp310-cp310-manylinux2014_x86_64.whl" \
+    ml-collections==0.1.0 \
+    numpy==1.24.3 \
+    pandas==2.0.3 \
+    scipy==1.11.1 \
+    tensorflow-cpu==2.13.0 \
+    matplotlib
```
[the remark this answers is not in the document]
Comment (Jin-HoMLee) [reply docker/Dockerfile.pipeline]: Fixed in dafe40d — pinned to `matplotlib==3.10.8`, verified from the working Docker image on the production GPU VM.

File: docker/Dockerfile.pipeline [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+WORKDIR /opt/TCRdock
+
+# Download AlphaFold params (model_2_ptm only) and BLAST
+RUN mkdir -p alphafold_params/params && \
+    wget -q -O alphafold_params/params/params_model_2_ptm.npz \
+        "https://www.dropbox.com/s/e3uz9mwxkmmv35z/params_model_2_ptm.npz" && \
+    wget -q -O alphafold_params/params/tcrpmhc_run4_af_mhc_params_891.pkl \
+        "https://www.dropbox.com/s/jph8v1mfni1q4y8/tcrpmhc_run4_af_mhc_params_891.pkl" && \
+    python download_blast.py
+
```
[the remark this answers is not in the document]
Comment (Jin-HoMLee) [reply docker/Dockerfile.pipeline]: Fixed in 47f88e3 — added `?dl=1` to Dropbox URLs to force direct download, plus a post-download validation that checks file size (>1KB) and rejects HTML responses.

File: scripts/setup_cloud.sh [hunk 9]
```diff
@@ -172,7 +172,7 @@ echo "  2. Activate the Snakemake environment:"
 echo "       conda activate snakemake"
 echo "  3. From inside a tmux session, run the pipeline:"
 echo "       cd $REPO_DIR"
-echo "       snakemake --cores \$(nproc) --use-conda 2>&1 | tee pipeline.log ; bash auto_stop.sh"
+echo "       snakemake --cores \$(nproc) --use-conda 2>&1 | tee pipeline.log"
 echo ""
 echo "     The VM will shut down automatically when the pipeline finishes."
```
[the remark this answers is not in the document]
Comment (Jin-HoMLee) [reply scripts/setup_cloud.sh]: Fixed in 45cecac — clarified that the manual setup path does not auto-stop the VM. The automated `run_cloud_gpu.sh` workflow does handle shutdown (CPU VM stopped after upload, GPU VM shuts down after TCRdock).

File: workflow/scripts/generate_report.py [hunk 10]
```diff
@@ -254,12 +363,63 @@ def _df_to_html(df: pd.DataFrame, max_rows: int = 100) -> str:
 # Report generation
 # ---------------------------------------------------------------------------
 
+def _build_structure_section(pdb_path: Path, pred_df: pd.DataFrame) -> str:
+    """Build the Mol* 3D viewer section for the top TCRdock candidate.
+
+    The PDB is inlined as a JSON string so the HTML report is fully
+    self-contained and does not depend on local file paths at view time.
```
[the remark this answers is not in the document]
Comment (Jin-HoMLee) [reply workflow/scripts/generate_report.py:367]: Fixed in beccaa8 — added 11 unit tests for `_extract_chain_ids`, `_build_chain_legend`, and `_build_compnd_records` covering normal paths, edge cases, and PDB format constraints. All 15 tests pass.

File: workflow/scripts/run_tcrdock.py [hunk 11]
```diff
@@ -0,0 +1,494 @@
...
+# ---------------------------------------------------------------------------
+# Candidate selection
+# ---------------------------------------------------------------------------
+
+def select_top_candidates(
+    predictions_tsv: str | Path,
+    n_candidates: int = 1,
+) -> pd.DataFrame:
+    """Select the top N strong binders by IC50 from MHCflurry predictions.
+
```
[the remark this answers is not in the document]
Comment (Jin-HoMLee) [reply workflow/scripts/run_tcrdock.py:74]: Fixed in 3337930 — added 7 unit tests for `select_top_candidates` (strong/weak/no-binder/n_candidates) and `relabel_pdb_chains` (2-chain, 4-chain TCR-pMHC, empty). All pass.

File: README.md [hunk 12]
```diff
@@ -557,6 +606,11 @@ All parameters are in `config/config.yaml`.  Key options:
 | `assembly.upstream_nt` | `26` | Nucleotides upstream of junction |
 | `assembly.downstream_nt` | `24` | Nucleotides downstream of junction |
 | `filtering.min_normal_reads` | `2` | Min reads in normal to trust a junction as patient-specific |
+| `tcrdock.enabled` | `false` | Enable TCRdock structural validation (GPU required) |
+| `tcrdock.docker_image` | `tcrdock:latest` | Docker image built by `setup_tcrdock_vm.sh` |
+| `tcrdock.n_candidates` | `1` | Number of top candidates to model |
+| `tcrdock.fallback_hla` | `HLA-A*02:01` | HLA allele for TCRdock (until HLA typing #23) |
```
[the remark this answers is not in the document]
Comment (Jin-HoMLee) [reply README.md]: Fixed in 3477a3e — updated to show the actual A/B/C dict schema matching `config/tcrdock_gpu.yaml`.

File: workflow/rules/tcrdock.smk [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+# =============================================================================
+#
+# Predicts the 3D structure of the TCR-peptide-MHC ternary complex for the
+# top-ranked neoepitope candidate(s).
+#
+# Requirements
+# ------------
+#   - Linux x64 with NVIDIA GPU (AlphaFold v2 backend)
+#   - TCRdock installed at config[tcrdock][tcrdock_dir]
+#   - AlphaFold params at config[tcrdock][alphafold_params_dir]
```
[the remark this answers is not in the document]
Comment (Jin-HoMLee) [reply workflow/rules/tcrdock.smk]: Fixed in 26b1b0d — updated to reference Docker image at `config[tcrdock][docker_image]`.

File: workflow/rules/tcrdock.smk [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+#
+# This rule is skipped entirely when config[tcrdock][enabled] is false,
+# so local / macOS arm64 / CPU-only runs are unaffected.
+#
+# Outputs
+# -------
+#   structures/{cancer_type}/top_candidate.pdb   — predicted ternary complex
+#   structures/{cancer_type}/docking_scores.tsv  — docking geometry metrics
```
[the remark this answers is not in the document]
Comment (Jin-HoMLee) [reply workflow/rules/tcrdock.smk]: Fixed in 26b1b0d — output paths now match the actual rule outputs under `results/predictions/{cancer_type}/tcrdock/`.

File: docs/google_cloud_guide.md [hunk 15]
```diff
@@ -291,21 +292,14 @@ tmux new -s pipeline
 snakemake --cores 4 --use-conda -n
 
 # Inside tmux: full run — pipeline logs to pipeline.log, VM shuts down when done
-snakemake --cores $(nproc) --use-conda --rerun-triggers mtime 2>&1 | tee pipeline.log ; bash auto_stop.sh
+snakemake --cores $(nproc) --use-conda --rerun-triggers mtime 2>&1 | tee pipeline.log
 ```
```
[the remark this answers is not in the document]
Comment (Jin-HoMLee) [reply docs/google_cloud_guide.md]: Fixed in ad8c859 — removed false auto-shutdown claim, added VM stop reminder and callout to `run_cloud_gpu.sh` for the automated workflow.

File: docs/google_cloud_guide.md [hunk 16]
```diff
@@ -567,7 +630,7 @@ tumor_01	Primary Tumor	data/tumor_01_R1.fastq	data/tumor_01_R2.fastq
 EOF
 
 # 8. Run pipeline (auto-shuts down VM when done to save costs)
```
[the remark this answers is not in the document]
Comment (Jin-HoMLee) [reply docs/google_cloud_guide.md]: Fixed in ad8c859 — updated the full example to clarify no auto-shutdown, added `run_cloud_gpu.sh --detach` as the automated alternative. Also updated GCS bucket references throughout the guide to use the project-derived name.

**Comment by Jin-HoMLee** [comment]
I am doing a last test run on cloud before merging...
